File: ontrack-ui/src/main/java/net/nemerosa/ontrack/boot/graphql/schema/GQLRootQueryBuilds.java
package net.nemerosa.ontrack.boot.graphql.schema;

import graphql.schema.DataFetcher;
import graphql.schema.GraphQLFieldDefinition;
import net.nemerosa.ontrack.boot.graphql.support.GraphqlUtils;
import net.nemerosa.ontrack.model.buildfilter.BuildFilterProviderData;
import net.nemerosa.ontrack.model.exceptions.BranchNotFoundException;
import net.nemerosa.ontrack.model.exceptions.ProjectNotFoundException;
import net.nemerosa.ontrack.model.structure.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.Optional;

import static graphql.Scalars.GraphQLInt;
import static graphql.Scalars.GraphQLString;
import static graphql.schema.GraphQLArgument.newArgument;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
import static net.nemerosa.ontrack.boot.graphql.support.GraphqlUtils.checkArgList;
import static net.nemerosa.ontrack.boot.graphql.support.GraphqlUtils.stdList;

@Component
public class GQLRootQueryBuilds implements GQLRootQuery {

    public static final String PROJECT_ARGUMENT = "project";
    public static final String BRANCH_ARGUMENT = "branch";
    public static final String BUILD_BRANCH_FILTER_ARGUMENT = "buildBranchFilter";
    public static final String BUILD_PROJECT_FILTER_ARGUMENT = "buildProjectFilter";

    private final StructureService structureService;
    private final GQLTypeBuild build;
    private final GQLInputBuildStandardFilter inputBuildStandardFilter;
    private final GQLInputBuildSearchForm inputBuildSearchForm;

    @Autowired
    public GQLRootQueryBuilds(
            StructureService structureService,
            GQLTypeBuild build,
            GQLInputBuildStandardFilter inputBuildStandardFilter,
            GQLInputBuildSearchForm inputBuildSearchForm
    ) {
        this.structureService = structureService;
        this.build = build;
        this.inputBuildStandardFilter = inputBuildStandardFilter;
        this.inputBuildSearchForm = inputBuildSearchForm;
    }

    @Override
    public GraphQLFieldDefinition getFieldDefinition() {
        return newFieldDefinition()
                .name("builds")
                .type(stdList(build.getType()))
                .argument(
                        newArgument()
                                .name("id")
                                .description("ID of the build to look for")
                                .type(GraphQLInt)
                                .build()
                )
                .argument(
                        newArgument()
                                .name(PROJECT_ARGUMENT)
                                .description("Name of a project")
                                .type(GraphQLString)
                                .build()
                )
                .argument(
                        newArgument()
                                .name(BRANCH_ARGUMENT)
                                .description("Name of a branch - requires 'project' to be filled as well")
                                .type(GraphQLString)
                                .build()
                )
                .argument(
                        newArgument()
                                .name(BUILD_BRANCH_FILTER_ARGUMENT)
                                .description("Filter to apply for the builds on the branch - requires 'branch' to be filled.")
                                .type(inputBuildStandardFilter.getInputType())
                                .build()
                )
                .argument(
                        newArgument()
                                .name(BUILD_PROJECT_FILTER_ARGUMENT)
                                .description("Filter to apply for the builds on the project - requires 'project' to be filled.")
                                .type(inputBuildSearchForm.getInputType())
                                .build()
                )
                .dataFetcher(buildFetcher())
                .build();
    }

    private DataFetcher buildFetcher() {
        return environment -> {
            Integer id = environment.getArgument("id");
            Optional<String> oProject = GraphqlUtils.getStringArgument(environment, PROJECT_ARGUMENT);
            Optional<String> oBranch = GraphqlUtils.getStringArgument(environment, BRANCH_ARGUMENT);
            Object branchFilter = environment.getArgument(BUILD_BRANCH_FILTER_ARGUMENT);
            Object projectFilter = environment.getArgument(BUILD_PROJECT_FILTER_ARGUMENT);
            // Per ID
            if (id != null) {
                checkArgList(environment, "id");
                return Collections.singletonList(
                        structureService.getBuild(ID.of(id))
                );
            }
            // Per project
            else if (oProject.isPresent()) {
                // ... and branch
                if (oBranch.isPresent()) {
                    // Gets the branch
                    Branch branch = structureService.findBranchByName(oProject.get(), oBranch.get())
                            .orElseThrow(() -> new BranchNotFoundException(oProject.get(), oBranch.get()));
                    // Configurable branch filter
                    BuildFilterProviderData<?> filter = inputBuildStandardFilter.convert(branchFilter);
                    // Runs the filter
                    return filter.filterBranchBuilds(branch);
                }
                // Project only
                else {
                    // Gets the project
                    Project project = structureService.findProjectByName(oProject.get())
                            .orElseThrow(() -> new ProjectNotFoundException(oProject.get()));
                    // Build search form as argument
                    BuildSearchForm form = inputBuildSearchForm.convert(projectFilter);
                    return structureService.buildSearch(project.getId(), form);
                }
            }
            // Branch filter only - not accepted
            else if (branchFilter != null) {
                throw new IllegalStateException(String.format(
                        "%s must be used together with %s",
                        BUILD_BRANCH_FILTER_ARGUMENT,
                        BRANCH_ARGUMENT
                ));
            }
            // Project filter only - not accepted
            else if (projectFilter != null) {
                throw new IllegalStateException(String.format(
                        "%s must be used together with %s",
                        BUILD_PROJECT_FILTER_ARGUMENT,
                        PROJECT_ARGUMENT
                ));
            }
            // None
            else {
                return Collections.emptyList();
            }
        };
    }

}

File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLRootQueryBuilds.kt
package net.nemerosa.ontrack.graphql.schema

import graphql.Scalars
import graphql.schema.DataFetcher
import graphql.schema.DataFetchingEnvironment
import graphql.schema.GraphQLArgument
import graphql.schema.GraphQLFieldDefinition
import net.nemerosa.ontrack.common.getOrNull
import net.nemerosa.ontrack.graphql.exceptions.ArgumentMismatchException
import net.nemerosa.ontrack.graphql.support.listType
import net.nemerosa.ontrack.model.structure.Build
import net.nemerosa.ontrack.model.structure.ID.Companion.of
import net.nemerosa.ontrack.model.structure.StructureService
import org.springframework.stereotype.Component

@Component
class GQLRootQueryBuilds(
        private val structureService: StructureService,
        private val build: GQLTypeBuild,
        private val inputBuildStandardFilter: GQLInputBuildStandardFilter,
        private val inputBuildSearchForm: GQLInputBuildSearchForm
) : GQLRootQuery {

    override fun getFieldDefinition(): GraphQLFieldDefinition {
        return GraphQLFieldDefinition.newFieldDefinition()
                .name("builds")
                .type(listType(build.typeRef, nullable = true))
                .argument(
                        GraphQLArgument.newArgument()
                                .name("id")
                                .description("ID of the build to look for")
                                .type(Scalars.GraphQLInt)
                                .build()
                )
                .argument(
                        GraphQLArgument.newArgument()
                                .name(PROJECT_ARGUMENT)
                                .description("Name of a project")
                                .type(Scalars.GraphQLString)
                                .build()
                )
                .argument(
                        GraphQLArgument.newArgument()
                                .name(BRANCH_ARGUMENT)
                                .description("Name of a branch - requires 'project' to be filled as well")
                                .type(Scalars.GraphQLString)
                                .build()
                )
                .argument(
                        GraphQLArgument.newArgument()
                                .name(NAME_ARGUMENT)
                                .description("Name of a build - requires 'project' & 'branch' to be filled as well")
                                .type(Scalars.GraphQLString)
                                .build()
                )
                .argument(
                        GraphQLArgument.newArgument()
                                .name(BUILD_BRANCH_FILTER_ARGUMENT)
                                .description("Filter to apply for the builds on the branch - requires 'branch' to be filled.")
                                .type(inputBuildStandardFilter.typeRef)
                                .build()
                )
                .argument(
                        GraphQLArgument.newArgument()
                                .name(BUILD_PROJECT_FILTER_ARGUMENT)
                                .description("Filter to apply for the builds on the project - requires 'project' to be filled.")
                                .type(inputBuildSearchForm.typeRef)
                                .build()
                )
                .dataFetcher(buildFetcher())
                .build()
    }

    private fun buildFetcher(): DataFetcher<List<Build>> {
        return DataFetcher { environment: DataFetchingEnvironment ->
            val id = environment.getArgument<Int>("id")
            val projectName: String? = environment.getArgument(PROJECT_ARGUMENT)
            val branchName: String? = environment.getArgument(BRANCH_ARGUMENT)
            val name: String? = environment.getArgument(NAME_ARGUMENT)
            val branchFilter = environment.getArgument<Any>(BUILD_BRANCH_FILTER_ARGUMENT)
            val projectFilter = environment.getArgument<Any>(BUILD_PROJECT_FILTER_ARGUMENT)
            // Per ID
            if (id != null) {
                listOf<Build>(
                        structureService.getBuild(of(id))
                )
            } else if (!projectName.isNullOrBlank()) {
                // ... and branch
                if (!branchName.isNullOrBlank()) {
                    // Name only
                    if (!name.isNullOrBlank()) {
                        listOfNotNull(
                                structureService.findBuildByName(projectName, branchName, name)
                                        .getOrNull()
                        )
                    } else {
                        // Gets the branch
                        val branch = structureService.findBranchByName(projectName, branchName)
                                .getOrNull()
                                ?: return@DataFetcher emptyList()
                        // Runs the filter
                        val filter = inputBuildStandardFilter.convert(branchFilter)
                        filter.filterBranchBuilds(branch)
                    }
                } else {
                    // Gets the project
                    val (id1) = structureService.findProjectByName(projectName)
                            .getOrNull()
                            ?: return@DataFetcher emptyList()
                    // Build search form as argument
                    val form = inputBuildSearchForm.convert(projectFilter)
                    return@DataFetcher structureService.buildSearch(id1, form)
                }
            } else if (branchFilter != null) {
                throw ArgumentMismatchException("$BUILD_BRANCH_FILTER_ARGUMENT must be used together with $BRANCH_ARGUMENT")
            } else if (projectFilter != null) {
                throw ArgumentMismatchException("$BUILD_PROJECT_FILTER_ARGUMENT must be used together with $PROJECT_ARGUMENT")
            } else {
                emptyList<Build>()
            }
        }
    }

    companion object {
        const val PROJECT_ARGUMENT = "project"
        const val BRANCH_ARGUMENT = "branch"
        const val NAME_ARGUMENT = "name"
        const val BUILD_BRANCH_FILTER_ARGUMENT = "buildBranchFilter"
        const val BUILD_PROJECT_FILTER_ARGUMENT = "buildProjectFilter"
    }
}
